hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastTypeCoercion.java

```diff
@@ -0,0 +1,422 @@
+        if (types == null) {
+            return false;
+        }
+
+        // Do the coercion.
+
+        RelDataType commonType = types[types.length - 1];
+
+        boolean coerced = false;
+        for (int i = 0; i < types.length - 1; ++i) {
+            RelDataType type = types[i];
+            type = TYPE_FACTORY.createTypeWithNullability(commonType, type.isNullable());
+            boolean operandCoerced = coerceOperandType(binding.getScope(), binding.getCall(), i, type);
+            coerced |= operandCoerced;
+
+            // If the operand was coerced to integer type, reassign its CAST type
+            // back to the common type: '0':VARCHAR -> CAST('0' AS INT(31)):INT(0)
+            // -> CAST('0' AS INT(31)):INT(31).
+            if (operandCoerced && isInteger(type)) {
+                updateInferredType(binding.operand(i), type);
+            }
+        }
+
+        return coerced;
+    }
+
+    @Override
+    public boolean caseWhenCoercion(SqlCallBinding binding) {
+        // Infer types.
+
+        SqlCase call = (SqlCase) binding.getCall();
+        SqlNodeList thenOperands = call.getThenOperands();
+
+        List<SqlNode> operands = new ArrayList<>(thenOperands.size() + 1);
+        operands.addAll(thenOperands.getList());
+        operands.add(call.getElseOperand());
+
+        RelDataType[] types = inferTypes(binding.getScope(), operands, false);
+        if (types == null) {
+            return false;
+        }
+
+        // Do the coercion.
+
+        boolean coerced = false;
+        for (int i = 0; i < operands.size() - 1; ++i) {
+            coerced |= coerceElementType(binding.getScope(), thenOperands, i, types[i]);
+        }
+        coerced |= coerceOperandType(binding.getScope(), call, 3, types[operands.size() - 1]);
+
+        updateInferredType(call, types[types.length - 1]);
+
+        return coerced;
+    }
+
+    @Override
+    public RelDataType implicitCast(RelDataType in, SqlTypeFamily expected) {
+        // enables implicit conversion from CHAR to BOOLEAN
+        if (CHAR_TYPES.contains(typeName(in)) && expected == SqlTypeFamily.BOOLEAN) {
+            return TYPE_FACTORY.createSqlType(BOOLEAN, in.isNullable());
+        }
+
+        return super.implicitCast(in, expected);
+    }
+
+    @Override
+    protected void updateInferredType(SqlNode node, RelDataType type) {
+        ((HazelcastSqlValidator) validator).setKnownNodeType(node, type);
+        super.updateInferredType(node, type);
+    }
+
+    @Override
+    protected boolean coerceOperandType(SqlValidatorScope scope, SqlCall call, int index, RelDataType to) {
+        SqlNode operand = call.getOperandList().get(index);
+        RelDataType from = validator.deriveType(scope, operand);
+
+        // Just update the inferred type if casting is not needed. But if casting
+        // is not possible, still insert the cast to fail on its validation later.
```

Comment:
> But if casting is not possible, still insert the cast to fail on its validation later
Do we have a test for that? I am interested in the exact error that will be thrown in this case

Reply:
Looks like it was doing something useful at some point, I clearly remember I have added it to fix some test. But now the only difference I have found is error messages produced during processing of literals: "Cast function cannot convert value of type BOOLEAN to type TINYINT" vs "Literal 'FALSE' can not be parsed to type 'TINYINT'". It might make some difference for temporal types, initially I started writing tests for them too and later deleted the tests. I'm going to remove the `&& canCast(from, to)` part.